server/src/telemetry.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{str::FromStr, time::Duration};
+
+use http::{uri::PathAndQuery, HeaderMap, HeaderValue, Uri};
+use restate_core::{cancellation_watcher, TaskCenter, TaskKind};
+use restate_service_client::HttpClient;
+use restate_types::config::CommonOptions;
+use tokio::time::Instant;
+use tracing::debug;
+
+use crate::build_info::{
+    RESTATE_SERVER_DEBUG, RESTATE_SERVER_TARGET_TRIPLE, RESTATE_SERVER_VERSION,
+};
+
+static TELEMETRY_URI: &str = "https://restate.gateway.scarf.sh/restate-server/";
+const TELEMETRY_PERIOD: Duration = Duration::from_secs(3600 * 24);
+
+pub enum Telemetry {
+    Enabled(TelemetryEnabled),
+    Disabled,
+}
+
+pub struct TelemetryEnabled {
+    client: HttpClient,
+    session_id: String,
+    start_time: Instant,
+}
+
+impl Telemetry {
+    pub fn create(options: &CommonOptions) -> Self {
+        if options.disable_telemetry {
+            Self::Disabled
+        } else {
+            let client = HttpClient::from_options(&options.service_client.http);
+            let session_id = ulid::Ulid::new().to_string();
```

Comment:
So restarts of all kinds (e.g. also on failures) will be treated as new sessions. Does this mean that we might overcount the number of users? Would it make sense to persist the session id to disk to not double count restarts where the data is still present?

Reply:
yes it would, but scarf will determine 'unique sessions' based on client ip. in corporates with shared egress ips, this might batch together too many users into one unique session, hence the utility of the session id. yes, a semi permanent node id would have value, but its also arguably too much collection (eg, allowing us to correlate which ips a particular restate node pops up onto). the approach ive taken appears to be fairly standard

Reply:
Alright, overcounting seems also better than undercounting.